zebra-network/src/peer/connection.rs

```diff
@@ -859,93 +793,76 @@ where
+    async fn drive_peer_request(&mut self, req: Request) -> Result<(), PeerError> {
         trace!(?req);
         use tower::{load_shed::error::Overloaded, ServiceExt};
 
         if self.svc.ready_and().await.is_err() {
             // Treat all service readiness errors as Overloaded
             // TODO: treat `TryRecvError::Closed` in `Inbound::poll_ready` as a fatal error (#1655)
-            self.fail_with(PeerError::Overloaded);
-            return;
+            Err(PeerError::Overloaded)?
         }
 
         let rsp = match self.svc.call(req).await {
             Err(e) => {
                 if e.is::<Overloaded>() {
                     tracing::warn!("inbound service is overloaded, closing connection");
                     metrics::counter!("pool.closed.loadshed", 1);
-                    self.fail_with(PeerError::Overloaded);
+                    Err(PeerError::Overloaded)?
                 } else {
                     // We could send a reject to the remote peer, but that might cause
                     // them to disconnect, and we might be using them to sync blocks.
-                    // For similar reasons, we don't want to fail_with() here - we
-                    // only close the connection if the peer is doing something wrong.
+                    // For similar reasons, we don't return an error - we only close
+                    // the connection if the peer is doing something wrong.
                     error!(%e,
                            connection_state = ?self.state,
                            client_receiver = ?self.client_rx,
                            "error processing peer request");
                 }
-                return;
+                return Ok(());
             }
             Ok(rsp) => rsp,
         };
 
         match rsp {
             Response::Nil => { /* generic success, do nothing */ }
-            Response::Peers(addrs) => {
-                if let Err(e) = self.peer_tx.send(Message::Addr(addrs)).await {
-                    self.fail_with(e);
-                }
-            }
+            Response::Peers(addrs) => self.peer_tx.send(Message::Addr(addrs)).await?,
             Response::Transactions(transactions) => {
                 // Generate one tx message per transaction.
                 for transaction in transactions.into_iter() {
-                    if let Err(e) = self.peer_tx.send(Message::Tx(transaction)).await {
-                        self.fail_with(e);
-                    }
+                    self.peer_tx.send(Message::Tx(transaction)).await?;
                 }
             }
             Response::Blocks(blocks) => {
                 // Generate one block message per block.
                 for block in blocks.into_iter() {
-                    if let Err(e) = self.peer_tx.send(Message::Block(block)).await {
-                        self.fail_with(e);
-                    }
+                    self.peer_tx.send(Message::Block(block)).await?;
                 }
```

Comment:
Similarly, the early exit is probably correct here, and the loop probably caused those panics.